src/util/lp/vars_equivalence.h

```diff
@@ -0,0 +1,296 @@
+/*++
+  Copyright (c) 2017 Microsoft Corporation
+
+  Module Name:
+
+  <name>
+
+  Abstract:
+
+  <abstract>
+
+  Author:
+  Nikolaj Bjorner (nbjorner)
+  Lev Nachmanson (levnach)
+
+  Revision History:
+
+
+  --*/
+
+namespace nla {
+
+typedef lp::constraint_index     lpci;
+typedef lp::explanation          expl_set;
+typedef lp::var_index            lpvar;
+struct hash_svector {
+    size_t operator()(const unsigned_vector & v) const {
+        return svector_hash<unsigned_hash>()(v);
+    }
+};
+
+
+
+struct vars_equivalence {
+    
+    struct equiv {
+        lpvar                m_i;
+        lpvar                m_j;
+        rational             m_sign;
+        lpci                 m_c0;
+        lpci                 m_c1;
+
+        equiv(lpvar i, lpvar j, rational const& sign, lpci c0, lpci c1) :
+            m_i(i),
+            m_j(j),
+            m_sign(sign),
+            m_c0(c0),
+            m_c1(c1) {
+            SASSERT(m_i != m_j);
+        }
+    };
+
+    struct node {
+        unsigned            m_parent;   // points to m_equivs
+        svector<unsigned>   m_children;  // point to m_equivs
+        node() : m_parent(-1) {}
+    };
+   
+    //fields
+    
+    // The map from the variables to m_nodes
```

Comment:
use union-find data-structures with embedded doubly linked lists for equivalence class traversal?

Reply:
Postpone the revision of this class